[C++] Fix -Wnarrowing and -Woverflow due to signed bitfields on G++ ARM

anassinator opened the pull request:
Older versions of G++ (at least on ARM) complain about narrowing conversions and
overflows when setting the reference index (defined as an 11-bit signed integer
bitfield) to -1:

```
error: narrowing conversion of '-1' from 'int' to 'short unsigned int' inside { } [-Wnarrowing]
error: conversion from 'short unsigned int' to 'short unsigned int:11' changes value from '65535' to '2047'
```

This is due to the signedness of integer bitfields (not explicitly signed or
unsigned) being implementation-defined. This addresses this issue by explicitly
defining the signedness of all the bitfields in `flatbuffers::TypeCode` in C++.

aardappel commented:
I don't understand how this fixes things, since the types we had before also explicitly defined the signedness, and in fact are exactly what the typedef contains?

Almost seems like a compiler bug where it uses the type from the first element of the bitfield?

Though if this is necessary, maybe good to add a comment like `not using uint16_t here because older GCCs ..`

anassinator commented:
@aardappel I think my compiler treats `int16_t:11` as `uint16_t:11`. Only the `sequence_ref` field is having a problem. I know that for `int:N` it is definitely implementation-defined whether it is actually signed or not (see: https://en.cppreference.com/w/c/language/bit_field). I couldn't find a reference defining whether `intX_t:N` is also implementation-defined, so I'm assuming my compiler assumes it is. For those reasons (and to be consistent), I changed all the fields.

I added a comment as requested.

aardappel commented:
Ok, sounds good!
